code/atom.dm

```diff
@@ -908,20 +908,30 @@
 //same as above :)
 /atom/movable/setMaterial(var/datum/material/mat1, var/appearance = 1, var/setname = 1, var/copy = 1, var/use_descriptors = 0)
 	var/prev_mat_triggeronentered = (src.material && src.material.triggersOnEntered && src.material.triggersOnEntered.len)
+	var/prev_added_hasentered = src.material?.owner_hasentered_added
 	..(mat1,appearance,setname,copy,use_descriptors)
 	var/cur_mat_triggeronentered = (src.material && src.material.triggersOnEntered && src.material.triggersOnEntered.len)
+	src.material?.owner_hasentered_added = prev_added_hasentered
 
 	if (prev_mat_triggeronentered != cur_mat_triggeronentered)
 		if (isturf(src.loc))
-			if (!src.event_handler_flags & USE_HASENTERED)
-				if(cur_mat_triggeronentered)
+			if (~src.event_handler_flags & USE_HASENTERED)
```

Comment:
Double checking, you want ~ where it is?

Reply:
Nah, though technically correct it only serves to obfuscate intended behavior.